02_knight_and_princess/kirill_tolstobrov/main.cpp

```diff
@@ -0,0 +1,145 @@
+#include <fstream>
+#include <iostream>
+#include <iterator>
+#include <map>
+#include <utility>
+#include <vector>
+
+using VertexId = int;
+using EdgeId = int;
+
+class Vertex {
+ public:
+  explicit Vertex(const VertexId& init_id = 0) : id(init_id) {}
+  void add_edge(const EdgeId& edge_id) { connected_edges_.push_back(edge_id); }
+  operator std::string() {
+    std::string result = "";
+    result += "    {\n      \"id\": " + std::to_string(id);
+    result += ",\n      \"edge_ids\": [";
+
+    for (int i = 0; i < connected_edges_.size(); i++) {
+      result += std::to_string(connected_edges_[i]);
+      if (i == connected_edges_.size() - 1) {
+        result += "]\n";
+      } else {
+        result += ", ";
+      }
+    }
+    result += "\n    }";
+    return result;
+  }
+
+  const VertexId id;
+
+ private:
+  std::vector<EdgeId> connected_edges_;
+};
+
+class Edge {
+ public:
+  Edge(const EdgeId& init_id, const VertexId& v1_init, const VertexId& v2_init)
+      : id(init_id), vertex1_id(v1_init), vertex2_id(v2_init){};
+  operator std::string() {
+    std::string result = "";
+    result += "    {\n      \"id\": " + std::to_string(id);
+    result += ",\n      \"vertex_ids\": ";
+
+    result += "[" + std::to_string(vertex1_id);
+    result += ", " + std::to_string(vertex2_id);
+    result += "]\n    }";
+
+    return result;
+  }
+
+  const EdgeId id;
+  const VertexId vertex1_id;
+  const VertexId vertex2_id;
+};
+
+class Graph {
+ public:
+  int get_vertices_amount() const { return vertices_.size(); }
+  int get_edges_amount() const { return edges_.size(); }
+
+  void add_new_vertex() { vertices_.push_back(Vertex(vertices_.size())); }
+
+  void bind_vertices(const VertexId& id1, const VertexId& id2) {
+    const auto& vertex1_connections = connections_map_[id1];
+    for (const auto& connection : vertex1_connections) {
+      if (connection.second == id2) {
+        std::cout << "Attemptig to connect connected vertices: Error."
+                  << std::endl;
+        return;
+      }
+    }
```

Comment:
Так как у нас нет входных данных от юзера, или же из сети, то единственная причина писать валидацию - проверить что в коде нету багов, а не обработать ожидаемую ошибку. В таких случаях нужно использовать `assert()`, так как они удаляются из `production` кода и не несут в себе последующего оверхеда для конечного юзера.
```suggestion
    assert(!is_vertices_connected(id1, id2) && "Attempting to connect already connected vertices: Error.");
```